Source/Dafny/Translator.cs

```diff
@@ -94,6 +96,229 @@ public string FreshNumericId(string prefix = "")
+          return String.Format("({0} + {1})", left, right);
+        } else if (op == BinaryExpr.Opcode.Sub) {
+          return String.Format("({0} - {1})", left, right);
+        } else if (op == BinaryExpr.Opcode.Mul) {
+          return String.Format("({0} * {1})", left, right);
+        } else {
+          Console.WriteLine("unhandled expr: {0}", expr);
+          throw new cce.UnreachableException();
+        }
+      } else if (expr is NameSegment) {
+        var name = (expr as NameSegment).Name;
+        this.variables.Add(name);
+        return name;
+      } else if (expr is LiteralExpr) {
+        return String.Format("({0})", (expr as LiteralExpr).Value);
+      } else if (expr is NegationExpression) {
+        var sub = this.EncodeSubExpr((expr as NegationExpression).E);
+        return String.Format("(-{0})", sub);
+      } else if (expr is ParensExpression) {
+        var sub = this.EncodeSubExpr((expr as ParensExpression).E);
+        return String.Format("({0})", sub);
+      } else {
+        // other expressions are treated as opaque
+        // we offer limited support to create temp variables that represent their values
+        var sub = this.OpacifySubExpr(expr);
+        if (!this.opaqueExprs.ContainsKey(sub)) {
+          var temp = this.AllocateFreshTemp();
+          this.opaqueExprs.Add(sub, temp);
+        }
+        return this.opaqueExprs[sub];
+      }
+    }
+
+    string EncodeEqualityPoly(Expression e0, Expression e1) {
+      var left = this.EncodeSubExpr(e0);
+      var right = this.EncodeSubExpr(e1);
+      var poly = String.Format("{0} - {1}", left, right);
+      return poly;
+    }
+
+    void EncodeGeneratorPoly(Expression expr) {
+      if (expr is BinaryExpr) {
+        var binExpr = expr as BinaryExpr;
+        if (binExpr.Op != BinaryExpr.Opcode.Eq) {
+          Console.WriteLine("Error: Grobner generator expression should be an Equality or IsModEquivalent");
+          throw new cce.UnreachableException();
+        }
+        polys.Add(this.EncodeEqualityPoly(binExpr.E0, binExpr.E1));
+      } else if (expr is ApplySuffix) {
+        var appplyExpr = expr as ApplySuffix;
+        if (!(appplyExpr.Lhs is NameSegment) || !(appplyExpr.Lhs as NameSegment).Name.Equals("IsModEquivalent")) {
+          Console.WriteLine("Error: Grobner generator expression should be an Equality or IsModEquivalent");
+          throw new cce.UnreachableException();
+        }
+        var left = this.EncodeSubExpr(appplyExpr.Args[0]);
+        var right = this.EncodeSubExpr(appplyExpr.Args[1]);
+        var mod = this.EncodeSubExpr(appplyExpr.Args[2]);
+        var temp = this.AllocateFreshTemp();
+        var poly = String.Format("{0} - {1} - {2} * {3}", left, right, mod, temp);
+        polys.Add(poly);
+      } else {
+        Console.WriteLine("unhandled generator expr: {0}", expr);
+        throw new cce.UnreachableException();
+      }
+    }
+
+    string EncodeQueryPoly(Expression expr) {
```

Comment:
This looks very similar to `EncodeGeneratorPoly`.  Would it make sense to factor out the common bits?

Reply:
That is true, I will try extract some common bits.